<a href="https://colab.research.google.com/github/Prashkov1ch/python-ai-Prashkovich-Anna/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_Lab01_Economists_Student_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Лабораторная работа 1 (экономисты): Регулярные выражения и «массивы» как списки

**Ограничения:** используйте только стандартную библиотеку Python и модуль `re`.  
Массивы в этой работе — обычные **списки** (`list`). Никаких `numpy`, `pandas` и т.п.

**Данные:** используются Excel-файлы бухгалтерской отчётности компаний (10 файлов).  
В этой лабораторной мы читаем отдельные текстовые поля (ИНН, ОКПО, ОКВЭД, адрес, наименование) и применяем к ним regex.

In [6]:
import re
from openpyxl import load_workbook
import os
from google.colab import drive
drive.mount('/content/drive')
PATH = '/content/drive/MyDrive/цифровая кафедра/'  # Лабораторные_Экономисты замените на имя своей папки
os.chdir(PATH)
FILES = ['Копия 1 Атомэнергопром.xlsx', 'Копия 2 Аэрофлот.xlsx', 'Копия 3 Газпром_петрозаводск.xlsx', 'Копия 4 Лукойл.xlsx', 'Копия 5 Роснефть.xlsx', 'Копия 6 Самолет.xlsx', 'Копия 7 Славмо.xlsx', 'Копия 8 Строительная_компания_Век.xlsx', 'Копия 9 ТГК_1.xlsx', 'Копия 10 ТНС_ЭНЭРГО_Карелия.xlsx']

def load_org_info(xlsx_name: str):
    wb = load_workbook(xlsx_name, data_only=True)
    ws = wb["Сведения об организации"]
    # В этих файлах значения для ключевых полей находятся в колонке 13 (M)
    def v(row, col=13):
        return ws.cell(row=row, column=col).value
    return {
        "file": xlsx_name,
        "full_name": v(6),
        "inn": v(10),
        "kpp": v(11),
        "okpo": v(12),
        "okved": v(15),
        "address": v(16),
        "ogrn": v(21),
    }

org_list = [load_org_info(fn) for fn in FILES]

# Список строк "код наименование" из отчёта о фин. результатах (для regex-упражнений)
def load_fin_lines(xlsx_name: str, limit_rows=120):
    wb = load_workbook(xlsx_name, data_only=True)
    ws = wb["Отчет о финансовых результатах"]
    lines = []
    for r in range(6, limit_rows+1):
        code = ws.cell(row=r, column=16).value  # "Код строки"
        name = ws.cell(row=r, column=5).value   # "Наименование показателя"
        if code is None or name is None:
            continue
        s = f"{str(code).strip()} {str(name).strip()}"
        if re.fullmatch(r"\d{4}", str(code).strip()):
            lines.append(s)
    return lines

fin_lines = load_fin_lines(FILES[0])  # для упражнений достаточно одного файла

print("Загружено компаний:", len(org_list))
print("Пример записи:", org_list[0])
print("Пример строк отчета о фин.результатах:", fin_lines[:5])

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/usr/local/lib/python3.12/dist-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


Загружено компаний: 10
Пример записи: {'file': 'Копия 1 Атомэнергопром.xlsx', 'full_name': 'АКЦИОНЕРНОЕ ОБЩЕСТВО "АТОМНЫЙ ЭНЕРГОПРОМЫШЛЕННЫЙ КОМПЛЕКС"', 'inn': '7706664260', 'kpp': '770601001', 'okpo': '81527864', 'okved': '62.09', 'address': '119017, Москва г, ул Ордынка Б., 24', 'ogrn': '1027700058286'}
Пример строк отчета о фин.результатах: ['2110 Выручка4', '2120 Себестоимость продаж', '2100 Валовая прибыль (убыток)', '2210 Коммерческие расходы', '2220 Управленческие расходы']


### Задание 1
Используя `re`, разберите имена файлов вида `"<номер> <название>.xlsx"`.  
Сформируйте список `companies` из кортежей `(номер:int, название:str)` и отсортируйте по номеру.

### Задание 2
Проверьте ИНН компаний регулярным выражением. Для юр. лиц ожидается ровно **10 цифр**.  
Выведите список файлов, где ИНН не прошёл проверку.

In [ ]:
# Ваш код здесь

### Задание 3
Проверьте ОКПО: ожидается ровно **8 цифр**.  
Выведите количество корректных и некорректных значений.

In [ ]:
# Ваш код здесь

### Задание 4
Из ОКВЭД (пример: `62.09`) извлеките две группы цифр: `section` и `group`.  
Сформируйте список словарей `okved_parts = [{"file":..., "section":.., "group":..}, ...]`.

In [ ]:
# Ваш код здесь

### Задание 5
Подсчитайте частоты `section` из ОКВЭД (используйте только списки и dict, без pandas).  
Результат: словарь `freq`, где ключ — section, значение — количество компаний.

In [ ]:
# Ваш код здесь

### Задание 6
Из адреса извлеките почтовый индекс (6 цифр) и город (слово перед `г,`).  
Выведите 5 примеров в формате: `файл -> индекс, город`.

In [ ]:
# Ваш код здесь

### Задание 7
Нормализуйте полное наименование компании:  
1) удалите кавычки `"`; 2) замените множественные пробелы на один; 3) уберите пробелы по краям.  
Сформируйте список `clean_names` длины 10.

In [ ]:
# Ваш код здесь

### Задание 8
Разбейте нормализованное название компании на токены с помощью `re.split`,  
где разделители: пробелы, запятые, точки. Покажите первые 10 токенов для первой компании.

In [ ]:
# Ваш код здесь

### Задание 9
Используя `re.finditer`, извлеките из списка `fin_lines` все пары (код, наименование).  
Код — 4 цифры в начале строки. Сформируйте словарь `fin_map[code]=name`.

In [ ]:
# Ваш код здесь

### Задание 10
В строках `fin_lines` найдите все упоминания слова «прибыль» в любом регистре (IGNORECASE).  
Выведите список строк, где найдено совпадение.

In [ ]:
# Ваш код здесь

### Задание 11
Преобразуйте ИНН в формат `ИНН: 7706 664260` (пример: 4 цифры, пробел, 6 цифр).  
Используйте `re.sub` с группами. Выведите 5 примеров.

In [ ]:
# Ваш код здесь

### Задание 12
Определите организационно-правовую форму по названию:  
- если есть «ПУБЛИЧНОЕ АКЦИОНЕРНОЕ ОБЩЕСТВО» → `PAO`  
- иначе если есть «АКЦИОНЕРНОЕ ОБЩЕСТВО» → `AO`  
- иначе → `OTHER`  
Используйте `re.search` с флагом IGNORECASE. Выведите пары: `файл -> форма`.

In [ ]:
# Ваш код здесь